### PR TITLE
Bank of Corda demo - Issuer of Cash

### PR DESCRIPTION
**Pull request by josecoll**
Resolve BankOfCorda through NMS in protocol

Fixes following Integration testing.

Register custom RPC Kryo classes.

Protocol -> Flow renaming

Bank of Corda demo - Issuer of Cash

Resolve BankOfCorda through NMS in protocol

Fixes following Integration testing.

Protocol -> Flow renaming

Addressed all comments in PR review.

Removed bank lines.

Updated minor inconsistency in README.md

All protocol references changed to flow.

changed protocol -> flow in TODO comment.

changed startProtocolPermission -> startFlowPermission in README.md

Added transaction id to IssuerFlow Success response.
Removed explicit call to record Cash Move transaction (as already recorded in subflow)

Removed quasar dependency.

Addressed comment in PR.

Updated to use CompositeKey.

Added arguments to pass in Currency and Amount.

Updated run configurations to pass in Currency and Amount values

Added additional parameter to IssuerFlow request: issueToPartyReference

Added Vault updates verification in RPC Integration test.

Fixed RPC Integration test (Vault assertions)

### CODE REVIEW

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/api/BankOfCordaClientApi.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        client.start("user1","test")
+        val proxy = client.proxy()
+
+        val amount = Amount(params.amount, currency(params.currency))
+        val issuerToPartyRef = OpaqueBytes.of(params.issueToPartyRefAsString.toByte())
+        return proxy.startFlow(::IssuanceRequester, amount, params.issueToPartyName, issuerToPartyRef, BOC_ISSUER_PARTY.name).returnValue.toBlocking().first()
+    }
+
+    private companion object {
+        private val apiRoot = "api/bank"
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/api/BankOfCordaClientApi.kt]: You can inline this - it's only used in one place (and can only be used in one place, as it's private) - so making it a named constant doesn't improve anything.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import net.corda.core.serialization.OpaqueBytes
+import net.corda.core.utilities.ProgressTracker
+import net.corda.flows.CashCommand
+import net.corda.flows.CashFlow
+import net.corda.flows.CashFlowResult
+import java.util.*
+
+/**
+ *  This flow enables a client to request issuance of some [FungibleAsset] from a
+ *  server acting as an issuer (see [Issued]) of FungibleAssets
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: Can you add some verbiage about how this isn't useful in real settings, and in fact should not be enabled, but it's useful as a source of fake assets for testing and experimentation.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ */
+object IssuerFlow {
+    data class IssuanceRequestState(val amount: Amount<Currency>, val issueToParty: Party, val issuerPartyRef: OpaqueBytes)
+
+    /*
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: /**

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                return IssuerFlowResult.Failed("Unable to locate ${otherParty} in Network Map Service")
+            }
+            else {
+                val issueRequest = IssuanceRequestState(amount, issueToParty, issueToPartyRef)
+                return sendAndReceive<IssuerFlowResult>(bankOfCordaParty, issueRequest).unwrap { it }
+            }
+        }
+    }
+
+    /*
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: /**

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ */
+object IssuerFlow {
+    data class IssuanceRequestState(val amount: Amount<Currency>, val issueToParty: Party, val issuerPartyRef: OpaqueBytes)
+
+    /*
+     * IssuanceRequester refers to a Node acting as issuance requester of some FungibleAsset
+     */
+    class IssuanceRequester(val amount: Amount<Currency>, val issueToPartyName: String, val issueToPartyRef: OpaqueBytes,
+                            val otherParty: String): FlowLogic<IssuerFlowResult>() {
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: It's best to keep APIs as type safe as possible - if someone finds themselves in a position where they have a string but they want to use `IssuanceRequester` then they can do the lookup themselves. But they may already have a `Party` object in which case round-tripping is wasteful and can lead to bugs. So just push the lookup to the users here.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+/**
+ *  This flow enables a client to request issuance of some [FungibleAsset] from a
+ *  server acting as an issuer (see [Issued]) of FungibleAssets
+ *
+ */
+object IssuerFlow {
+    data class IssuanceRequestState(val amount: Amount<Currency>, val issueToParty: Party, val issuerPartyRef: OpaqueBytes)
+
+    /*
+     * IssuanceRequester refers to a Node acting as issuance requester of some FungibleAsset
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: This comment doesn't make a whole lot of sense to my ear and at any rate, seems to duplicate the story already told by the code itself:

https://docs.corda.net/codestyle.html#comments

Maybe something like this would be more helpful:

> This flow asks a remote node to issue us some cash with the given details. It may reject that request.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        @Suspendable
+        override fun call(): IssuerFlowResult {
+            val issueToParty = serviceHub.identityService.partyFromName(issueToPartyName)
+            val bankOfCordaParty = serviceHub.identityService.partyFromName(otherParty)
+            if (issueToParty == null || bankOfCordaParty == null) {
+                return IssuerFlowResult.Failed("Unable to locate ${otherParty} in Network Map Service")
+            }
+            else {
+                val issueRequest = IssuanceRequestState(amount, issueToParty, issueToPartyRef)
+                return sendAndReceive<IssuerFlowResult>(bankOfCordaParty, issueRequest).unwrap { it }
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: I know it's not implemented yet but error handling in flows is meant to be exception based. Can you just do the `sendAndReceive` here with the return type being Unit?

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            }
+        }
+    }
+
+    /*
+     * Issuer refers to a Node acting as a Bank Issuer of FungibleAssets
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: Same thing: this class doc could be written in a clearer way, I'm sure of it.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+     */
+    class Issuer(val otherParty: Party,
+                 override val progressTracker: ProgressTracker = Issuer.tracker()): FlowLogic<IssuerFlowResult>() {
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: The progressTracker parameter here is never actually replaced anywhere with anything, nor would it really make sense do to so. If you move `override val progressTracker: ProgressTracker = tracker()` into the class body then you can write the registration below in a simpler way.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        companion object {
+            object AWAITING_REQUEST : ProgressTracker.Step("Awaiting issuance request")
+            object ISSUING : ProgressTracker.Step("Self issuing asset")
+            object TRANSFERRING : ProgressTracker.Step("Transferring asset to issuance requester")
+            object SENDING_CONIFIRM : ProgressTracker.Step("Confirming asset issuance to requester")
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: typo: SENDING_CONFIRM

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            fun tracker() = ProgressTracker(AWAITING_REQUEST, ISSUING, TRANSFERRING, SENDING_CONIFIRM)
+        }
+
+        @Suspendable
+        override fun call(): IssuerFlowResult {
+            progressTracker.currentStep = AWAITING_REQUEST
+            val issueRequest = receive<IssuanceRequestState>(otherParty).unwrap { it }
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: Are there really no validation conditions on the request? Maybe we should restrict the currencies we're willing to issue?

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            if (resultMove is CashFlowResult.Failed) {
+                logger.error("Problem transferring cash: ${resultMove.message}")
+                return resultMove
+            }
+            return resultMove
+        }
+
+        class Service(services: PluginServiceHub) {
+            init {
+                services.registerFlowInitiator(IssuanceRequester::class) {
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: If you move progressTracker, you can write this like so:

`services.registerFlowInitiator(IssuanceRequester::class, Issuer::class)`

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                    Issuer(it)
+                }
+            }
+        }
+    }
+}
+
+sealed class IssuerFlowResult {
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: If you switch to just using exceptions, this class hierarchy is no longer necessary.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/plugin/BankOfCordaPlugin.kt [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    // A list of classes that expose web APIs.
+    override val webApis: List<Class<*>> = listOf(BankOfCordaWebApi::class.java)
+    // A list of flow that are required for this cordapp
+    override val requiredFlows: Map<String, Set<String>> =
+        mapOf(IssuerFlow.IssuanceRequester::class.java.name to setOf(Amount::class.java.name, String::class.java.name, OpaqueBytes::class.java.name, String::class.java.name)
+    )
+    override val servicePlugins: List<Class<*>> = listOf(IssuerFlow.Issuer.Service::class.java)
+
+    override fun registerRPCKryoTypes(kryo: Kryo): Boolean {
+        kryo.register(IssuerFlowResult.Success::class.java)
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/plugin/BankOfCordaPlugin.kt]: ditto - RPC can handle exceptions if you register the exception types.

File: samples/bank-of-corda-demo/src/test/kotlin/net/corda/bank/flow/IssuerFlowTest.kt [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+import net.corda.testing.MEGA_CORP_KEY
+import net.corda.testing.initiateSingleShotFlow
+import net.corda.testing.ledger
+import net.corda.testing.node.MockNetwork
+import org.junit.Test
+import java.util.*
+import kotlin.test.assertEquals
+
+class IssuerFlowTest {
+
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/test/kotlin/net/corda/bank/flow/IssuerFlowTest.kt]: blank line here

File: samples/bank-of-corda-demo/src/test/kotlin/net/corda/bank/flow/IssuerFlowTest.kt [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    @Test
+    fun `test issuer flow`() {
+        net = MockNetwork(false, true)
+        ledger {
+            notaryNode = net.createNotaryNode(DUMMY_NOTARY.name, DUMMY_NOTARY_KEY)
+            bankOfCordaNode = net.createPartyNode(notaryNode.info.address, BOC_ISSUER_PARTY.name, BOC_KEY)
+            bankClientNode = net.createPartyNode(notaryNode.info.address, MEGA_CORP.name, MEGA_CORP_KEY)
+
+            bankClientNode.disableDBCloseOnStop()
```
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/test/kotlin/net/corda/bank/flow/IssuerFlowTest.kt]: Why is this needed, given that you are closing the dbs right after calling stop?

File: samples/bank-of-corda-demo/build.gradle [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+            artifact sourceJar
+            artifact javadocJar
+        }
+    }
+}
+
+task runIssuer(type: JavaExec) {
+    classpath = sourceSets.main.runtimeClasspath
+    main = 'et.corda.bank.BankOfCordaDriverKt'
```
Comment (josecoll) [inline samples/bank-of-corda-demo/build.gradle]: Odd - yes I did a complete run-through.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt (hunk 5 above)
Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: Ah, I see. So it's due to missing RPCs.

This would be a good opportunity to flesh out the RPC interface a bit more. Why not add identity lookups in another PR and then rebase this one onto it?

Comment (mikehearn) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: Ah, OK. But maybe there should be an easier to use RPC than loading the entire network map. Like one that can do fuzzy search too.

Comment (josecoll) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: IssueFlow API now expect Party types in request
Parties are resolved by client prior to calling:
1) RPC client resolves via RPC networkUpdates API
2) Web API resolves via ServiceHub Identity Service

**Comment by mikehearn** [comment]
I think you'll need to tweak the web API code too, as the web server will only have RPC access after Andras' PR is merged in.

File: samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt (hunk 12 above)
Comment (josecoll) [inline samples/bank-of-corda-demo/src/main/kotlin/net/corda/bank/flow/IssuerFlow.kt]: I cannot get this to work as suggested above???

**Comment by josecoll** [comment]
Completed changes to use new WebAPI RPC interface.
Please review and approve.

**Comment by mikehearn** [comment]
LGTM
